evennia/contrib/aisystem/README.md

```diff
@@ -0,0 +1,661 @@
+@aicopynode - attaches a copy of the specified node to another node
+
+@aimovenode - moves a specified node under another node
+
+@aiswap - swaps the locations of two nodes
+
+@aishift - shifts the position of a node in its parent node's list of children
+
+@airemove - removes a node from its tree
+
+Note that, at present, there is no way to convert AI trees stored in the
+game's database into Python code. I can make that feature available on
+request, however.
+
+### Adding your own node types
+
+If you are not a coder, this section will not be useful to you.
+
+The AI system provides barebones functionality, but you will want to create
+a set of nodes tailored to your own game. To this end, you need to subclass
+either one of the fully functional node classes that come with the AI system,
+or the LeafNode, DecoratorNode and CompositeNode classes, which have no
+functionality. You should never subclass your nodes from the Node or RootNode
+class.
+
+Each node you create has a `weight` property that should not be overriden, but
+can be assigned a float value to determine how likely this node is to be picked
+by a ProbSelector or ProbSequence node.
+
+Each node also has a `name` property that contains the name you assigned it,
+as well as a `hash` property, which contains three alphanumeric characters that
+are unique to the node within a given tree, followed by an underscore character
+and the id number of its tree. For instance: `H5q_1566`, `Zmr_43`, `7Fl_5425163`
+The first three characters of this hash are used by the AI system's various
+commands and can be used to reference the nodes when two or more nodes in the
+tree share the same name.
+
+Each node contains the following methods, all of which may be overridden.
+With the exception of the update() method, all of them should feature a call to
+`super()`. Many of the methods have the `bb` argument, which is the blackboard
+of the agent whose AI is running.
+
+`update(self, bb)` - this method performs the procedure that occurs whenever the
+node runs. If the node has a child, it should probably call it at some
+point via a line such as the following:
+
+```
+status = node.children.tick(bb)
+```
+
+Here, status is the child's return status, either SUCCESS, FAILURE, RUNNING or
+ERROR. As mentioned above in the "Setting up the AI system" section, all these
+statuses are constants that may be imported from the `evennia.contrib.aisystem`
+module.
+
+If the node is a composite node, it will have a list of children rather than
+one child. In this case, you will probably want to run some or all of its
+children. For instance:
+
+```
+for child in node.children: 
+    [...]
+    child.tick(bb)
+    [...]
+```
+
+Be sure to have the update method return a status, either SUCCESS, FAILURE,
+RUNNING or ERROR. Exceptions are automatically handled by the ai system (see
+below), but you can override the system's functionality.
+
+The tick method should always have one argument, the blackboard in the update
```

Comment:
It is still unclear to me as a reader just what kind of an object a blackboard is at this point ...